File: sdk/php/src/Patch.php
<?php
/**
 * NHTML Patch v0.4.0
 * Représente une collection de mutations DOM à envoyer au client.
 */

namespace Nhtml;

class Patch {
    protected array $ops = [];
    protected array $joinRooms = [];
    protected array $leaveRooms = [];
    protected ?array $broadcastInstr = null;

    public static function create(): self {
        return new static();
    }

    public function setText(string $nid, string $text): self {
        $this->ops[] = ['op' => 'set_text', 'nid' => $nid, 'val' => $text];
        return $this;
    }

    public function addClass(string $nid, string $class): self {
        $this->ops[] = ['op' => 'add_class', 'nid' => $nid, 'val' => $class];
        return $this;
    }

    public function removeClass(string $nid, string $class): self {
        $this->ops[] = ['op' => 'del_class', 'nid' => $nid, 'val' => $class];
        return $this;
    }

    public function setStyle(string $nid, string $prop, string $val): self {
        $this->ops[] = ['op' => 'set_style', 'nid' => $nid, 'prop' => $prop, 'val' => $val];
        return $this;
    }

    public function setAttr(string $nid, string $key, string $val): self {
        $this->ops[] = ['op' => 'set_attr', 'nid' => $nid, 'key' => $key, 'val' => $val];
        return $this;
    }

    public function delAttr(string $nid, string $key): self {
        $this->ops[] = ['op' => 'del_attr', 'nid' => $nid, 'key' => $key];
        return $this;
    }

    public function replaceInner(string $nid, string $html): self {
        $this->ops[] = ['op' => 'replace_inner', 'nid' => $nid, 'val' => $html];
        return $this;
    }

    public function appendHtml(string $nid, string $html): self {
        $this->ops[] = ['op' => 'append_html', 'nid' => $nid, 'val' => $html];
        return $this;
    }

    public function remove(string $nid): self {
        $this->ops[] = ['op' => 'remove', 'nid' => $nid];
        return $this;
    }

    public function focus(string $nid): self {
        $this->ops[] = ['op' => 'focus', 'nid' => $nid];
        return $this;
    }

    public function scrollTo(string $nid): self {
        $this->ops[] = ['op' => 'scroll_to', 'nid' => $nid];
        return $this;
    }

    /**
     * Marque la dernière opération comme devant être diffusée à tous les clients
     * @deprecated Utilisez broadcastInRoom() ou broadcastToOthers() pour plus de contrôle
     */
    public function broadcast(bool $b = true): self {
        if (!empty($this->ops)) {
            $this->ops[count($this->ops) - 1]['broadcast'] = $b;
        }
        return $this;
    }

    /**
     * Configure une diffusion groupée explicite (v0.6.0)
     */
    public function broadcastToAll(array $ops): self {
        $this->broadcastInstr = ['scope' => 'all', 'patch' => $ops];
        return $this;
    }

    public function broadcastToOthers(array $ops): self {
        $this->broadcastInstr = ['scope' => 'others', 'patch' => $ops];
        return $this;
    }

    public function broadcastInRoom(string $roomId, array $ops): self {
        $this->broadcastInstr = ['scope' => 'room', 'room_id' => $roomId, 'patch' => $ops];
        return $this;
    }

    public function broadcastToSession(string $sessionId, array $ops): self {
        $this->broadcastInstr = ['scope' => 'direct', 'target_sid' => $sessionId, 'patch' => $ops];
        return $this;
    }

    /**
     * Gestion des Salons
     */
    public function joinRoom(string $roomId): self {
        $this->joinRooms[] = $roomId;
        return $this;
    }

    public function leaveRoom(string $roomId): self {
        $this->leaveRooms[] = $roomId;
        return $this;
    }

    public function getOps(): array {
        return $this->ops;
    }

    /**
     * Envoie la réponse au Gateway Rust
     */
    public function send(): void {
        if (PHP_SAPI !== 'cli' && !headers_sent()) {
            header('Content-Type: application/json');
        }

        $response = [
            'patch' => $this->ops
        ];

        if (!empty($this->joinRooms)) {
            $response['join_room'] = $this->joinRooms;
        }

        if (!empty($this->leaveRooms)) {
            $response['leave_room'] = $this->leaveRooms;
        }

        if ($this->broadcastInstr) {
            $response['broadcast'] = $this->broadcastInstr;
        }

        echo json_encode($response);
        exit;
    }
}

File: sdk/go/patch.go
package nhtml

type PatchOp map[string]interface{}

type BroadcastInstruction struct {
	Scope     string    `json:"scope"`
	RoomID    string    `json:"room_id,omitempty"`
	TargetSID string    `json:"target_sid,omitempty"`
	Patch     []PatchOp `json:"patch"`
}

type Patch struct {
	Ops            []PatchOp             `json:"patch"`
	JoinRooms      []string              `json:"join_room,omitempty"`
	LeaveRooms     []string              `json:"leave_room,omitempty"`
	BroadcastInstr *BroadcastInstruction `json:"broadcast,omitempty"`
}

func NewPatch() *Patch {
	return &Patch{
		Ops:        []PatchOp{},
		JoinRooms:  []string{},
		LeaveRooms: []string{},
	}
}

func (p *Patch) SetText(nid string, text string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "set_text", "nid": nid, "val": text})
	return p
}

func (p *Patch) Broadcast(b bool) *Patch {
	if len(p.Ops) > 0 {
		p.Ops[len(p.Ops)-1]["broadcast"] = b
	}
	return p
}

func (p *Patch) AddClass(nid string, className string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "add_class", "nid": nid, "val": className})
	return p
}

func (p *Patch) RemoveClass(nid string, className string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "del_class", "nid": nid, "val": className})
	return p
}

func (p *Patch) SetStyle(nid string, prop string, val string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "set_style", "nid": nid, "prop": prop, "val": val})
	return p
}

func (p *Patch) SetAttr(nid string, key string, val string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "set_attr", "nid": nid, "key": key, "val": val})
	return p
}

func (p *Patch) DelAttr(nid string, key string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "del_attr", "nid": nid, "key": key})
	return p
}

func (p *Patch) ReplaceInner(nid string, html string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "replace_inner", "nid": nid, "val": html})
	return p
}

func (p *Patch) AppendHTML(nid string, html string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "append_html", "nid": nid, "val": html})
	return p
}

func (p *Patch) Remove(nid string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "remove", "nid": nid})
	return p
}

func (p *Patch) Focus(nid string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "focus", "nid": nid})
	return p
}

func (p *Patch) ScrollTo(nid string) *Patch {
	p.Ops = append(p.Ops, PatchOp{"op": "scroll_to", "nid": nid})
	return p
}

func (p *Patch) JoinRoom(roomID string) *Patch {
	p.JoinRooms = append(p.JoinRooms, roomID)
	return p
}

func (p *Patch) LeaveRoom(roomID string) *Patch {
	p.LeaveRooms = append(p.LeaveRooms, roomID)
	return p
}

func (p *Patch) BroadcastToAll(ops []PatchOp) *Patch {
	p.BroadcastInstr = &BroadcastInstruction{Scope: "all", Patch: ops}
	return p
}

func (p *Patch) BroadcastToOthers(ops []PatchOp) *Patch {
	p.BroadcastInstr = &BroadcastInstruction{Scope: "others", Patch: ops}
	return p
}

func (p *Patch) BroadcastInRoom(roomID string, ops []PatchOp) *Patch {
	p.BroadcastInstr = &BroadcastInstruction{Scope: "room", RoomID: roomID, Patch: ops}
	return p
}

func (p *Patch) BroadcastToSession(sessionID string, ops []PatchOp) *Patch {
	p.BroadcastInstr = &BroadcastInstruction{Scope: "direct", TargetSID: sessionID, Patch: ops}
	return p
}
